kong/dns/utils.lua

```diff
@@ -0,0 +1,285 @@
+        table_remove(resolv.search, i)
+      end
+    end
+  end
+
+  -- nameservers
+  if resolv.nameserver then
+    local nameservers = {}
+
+    for _, address in ipairs(resolv.nameserver) do
+      local ip, port, t = utils.parseHostname(address)
+      if t == "ipv4" or
+        (t == "ipv6" and not ip:find([[%]], nil, true) and enable_ipv6)
+      then
+        table_insert(nameservers, port and { ip, port } or ip)
+      end
+    end
+
+    resolv.nameservers = nameservers
+  end
+  return resolv
+end
+
+
+function _M.is_fqdn(name, ndots)
+  if name:sub(-1) == "." then
+    return true
+  end
+  local _, dot_count = name:gsub("%.", "")
+  return (dot_count >= ndots)
+end
+
+
+-- check if it matchs the SRV pattern: _<service>._<proto>.<name>
+function _M.is_srv(name)
+  return name:match("^_[^._]+%._[^._]+%.[^.]+") ~= nil
+end
+
+
+-- construct names from resolv options: search, ndots and domain
+function _M.search_names(name, resolv, hosts)
+  if not resolv.search or _M.is_fqdn(name, resolv.ndots) or
+    (hosts and hosts[name])
+  then
+    return { name }
+  end
+
+  local names = {}
+  for _, suffix in ipairs(resolv.search) do
+    table_insert(names, name .. "." .. suffix)
+  end
+
+  table_insert(names, name) -- append the original name at last
+
+  return names
+end
+
+
+-- add square brackets around IPv6 addresses if a non-strict check detects them
+function _M.ipv6_bracket(name)
+  if name:match("^[^[].*:") then  -- not start with '[' and contains ':'
+    return "[" .. name .. "]"
+  end
+  return name
+end
+
+
+-- util APIs to balance @answers
+
+function _M.get_next_round_robin_answer(answers)
+  answers.last = (answers.last or 0) % #answers + 1
+  return answers[answers.last]
+end
+
+
+do
+  -- based on the Nginx's SWRR algorithm and lua-resty-balancer
+  local function swrr_next(answers)
+    local total = 0
+    local best = nil    -- best answer in answers[]
+
+    for _, answer in ipairs(answers) do
+      -- 0.1 gives weight 0 record a minimal chance of being chosen (rfc 2782)
+      local w = (answer.weight == 0) and 0.1 or answer.weight
+      local cw = answer.cw + w
+      answer.cw = cw
+      if not best or cw > best.cw then
+        best = answer
+      end
+      total = total + w
+    end
+
+    best.cw = best.cw - total
+    return best
+  end
+
+
+  local function swrr_init(answers)
+    for _, answer in ipairs(answers) do
+      answer.cw = 0   -- current weight
+    end
+
+    -- random start
+    for _ = 1, math_random(#answers) do
+      swrr_next(answers)
+    end
+  end
+
+
+  -- gather records with the lowest priority in SRV record
+  local function filter_lowest_priority_answers(answers)
+    -- SRV record MUST have `priority` field
+    local lowest_priority = answers[1].priority
+    local l = {}    -- lowest priority records list
+
+    for _, answer in ipairs(answers) do
+      if answer.priority < lowest_priority then
+        lowest_priority = answer.priority
+        l = { answer }
```

Comment:
Let's clear the table `l` and assign the `answer` to `l[1]` to avoid creating too many tables.

Reply:
resolved

Reply:
https://github.com/Kong/kong/pull/12305/commits/f0d6e2a94395238c886a62eaa16710879e3f942b